common/archiver/s3store/visibility_archiver.go

```diff
@@ -209,12 +211,58 @@ func (v *visibilityArchiver) queryAll(
 	request *archiver.QueryVisibilityRequest,
 	saTypeMap searchattribute.NameTypeMap,
 ) (*archiver.QueryVisibilityResponse, error) {
-	return v.queryPrefix(ctx, uri, &queryVisibilityRequest{
-		namespaceID:   request.NamespaceID,
-		pageSize:      request.PageSize,
-		nextPageToken: request.NextPageToken,
-		parsedQuery:   &parsedQuery{},
-	}, saTypeMap, constructVisibilitySearchPrefix(uri.Path(), request.NamespaceID))
+	// remaining is the number of workflow executions left to return before we reach pageSize.
+	remaining := request.PageSize
+	nextPageToken := request.NextPageToken
+	var executions []*workflowpb.WorkflowExecutionInfo
+	// We need to loop because the number of workflow executions returned by each call to query may be fewer than
+	// pageSize. This is because we may have to skip some workflow executions after querying S3 (client-side filtering)
+	// because there are 2 entries in S3 for each workflow execution indexed by workflowTypeName (one for closeTimeout
+	// and one for startTimeout), and we only want to return one entry per workflow execution. See
+	// createIndexesToArchive for a list of all indexes.
+	for {
+		searchPrefix := constructVisibilitySearchPrefix(uri.Path(), request.NamespaceID)
+		// We suffix searchPrefix with workflowTypeName because the data in S3 is duplicated across combinations of 2
+		// different primary indices (workflowID and workflowTypeName) and 2 different secondary indices (closeTimeout
+		// and startTimeout). We only want to return one entry per workflow execution, but the full path to the S3 key
+		// is <primaryIndexKey>/<primaryIndexValue>/<secondaryIndexKey>/<secondaryIndexValue>/<runID>, and we don't have
+		// the primaryIndexValue when we make the call to query, so we can only specify the primaryIndexKey.
+		searchPrefix += "/" + primaryIndexKeyWorkflowTypeName
+		// The pageSize we supply here is actually the maximum number of keys to fetch from S3. For each execution,
+		// there should be 2 keys in S3 for this prefix, so you might think that we should multiply the pageSize by 2.
+		// However, if we do that, we may end up returning more than pageSize workflow executions to the end user of
+		// this API. This is because we aren't guaranteed that both keys for a given workflow execution will be returned
+		// in the same call. For example, if the user supplies a pageSize of 1, and we specify a maximum number of keys
+		// of 2 to S3, we may get back entries from S3 for 2 different workflow executions. You might think that we can
+		// just truncate this result to 1 workflow execution, but then the nextPageToken would be incorrect. So, we may
```

Comment:
btw, you probably considered this, but you could truncate and then modify the token to avoid this problem. or if you can't modify directly, wrap the token in a new token that has the old one plus some number of records to skip. it's probably more complexity than this is worth, though.